src/full/Agda/Syntax/Abstract/Pattern.hs

```diff
@@ -340,6 +340,32 @@ instance PatternToExpr Pattern Expr where
     WithP r p          -> __IMPOSSIBLE__
     AnnP _ _ p         -> patToExpr p
 
+-- | Make sure that there are no dot or equality patterns (called on pattern synonyms).
+--   Also disallows annotated patterns.
+--
+noDotOrEqPattern :: forall m e. Monad m
+  => m (A.Pattern' Void)   -- ^ Exception or replacement for dot (etc.) patterns.
+  -> A.Pattern' e          -- ^ In pattern.
+  -> m (A.Pattern' Void)   -- ^ Out pattern.
+noDotOrEqPattern err = dot
+  where
+    dot :: A.Pattern' e -> m (A.Pattern' Void)
+    dot = \case
+      A.VarP x               -> pure $ A.VarP x
+      A.ConP i c args        -> A.ConP i c <$> (traverse $ traverse $ traverse dot) args
+      A.ProjP i o d          -> pure $ A.ProjP i o d
+      A.WildP i              -> pure $ A.WildP i
+      A.AsP i x p            -> A.AsP i x <$> dot p
+      A.DotP{}               -> err
+      A.EqualP{}             -> err   -- Andrea: so we also disallow = patterns, reasonable?
+      A.AbsurdP i            -> pure $ A.AbsurdP i
+      A.LitP i l             -> pure $ A.LitP i l
+      A.DefP i f args        -> A.DefP i f <$> (traverse $ traverse $ traverse dot) args
```

Comment:
IIUC we should not have any `DefP` in here as they are only used in
display pragmas.

Outside the scope of this PR of course but I wonder whether it may
be worth turning `Pattern'` into a GADT to tighten the abstract syntax
representation. It feels like we have been throwing in extra constructors
out of convenience but now we have to handle impossible (?) cases in
a lot of places.

Reply:
Indeed, it has deteriorated over time.
Tighter fitting data structures would be better.

Vaguely related: When it comes to pattern representation, I do not even know why we have patterns separate from expressions in concrete syntax...